File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/graph/P797_AllPathsFromSourceToTarget.py
from typing import List


class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        res = []
        path = []
        self.traverse(graph, 0, path, res)
        return res

    def traverse(self, graph, idx, path, res):
        path.append(idx)
        if idx == len(graph) - 1:
            res.append(list(path))
            path.pop()
            return
        for i in graph[idx]:
            self.traverse(graph, i, path, res)
        path.pop()

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/graph/P797_AllPathsFromSourceToTarget.java
package com.dsa.leetcode.labuladong.graph;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @since 2024/6/9 19:11
 */
public class P797_AllPathsFromSourceToTarget {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        LinkedList<Integer> list = new LinkedList<>();
        traverse(graph, 0, list);
        return res;
    }

    public void traverse(int[][] graph, int rootIdx, LinkedList<Integer> list) {
        list.add(rootIdx);
        if (rootIdx == graph.length - 1) {
            res.add(new ArrayList<>(list));
            list.removeLast();
            return;
        }
        for (int i : graph[rootIdx]) {
            traverse(graph,i,list);
        }
        list.removeLast();
    }

}
